githubapp/client_creator.go

```diff
@@ -223,9 +240,34 @@ func (c *clientCreator) newV4Client(base *http.Client, details string) (*githubv
 	return client, nil
 }
 
-func applyMiddleware(base *http.Client, middleware []ClientMiddleware) {
+func (c *clientCreator) transport() http.RoundTripper {
+	if c.cacheFunc != nil {
+		return cacheControl(c.cacheControlEnabled)(httpcache.NewTransport(c.cacheFunc()))
+	}
+
+	return http.DefaultTransport
+}
+
+func applyMiddleware(base http.RoundTripper, middleware []ClientMiddleware) {
 	for i := len(middleware) - 1; i >= 0; i-- {
-		base.Transport = middleware[i](base.Transport)
+		base = middleware[i](base)
+	}
+}
+
+func cacheControl(enabled bool) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(r *http.Request) (*http.Response, error) {
+			resp, err := next.RoundTrip(r)
+
+			// Force validation to occur when the cache is disabled by setting max-age=0,
+			// as the cache results will always appear as stale
+			if !enabled {
+				cacheControl := resp.Header.Get("Cache-Control")
+				newCacheControl := maxAgeRegex.ReplaceAll([]byte(cacheControl), []byte("max-age=0"))
```

Comment:
I'm not incredibly happy about using a regex here, and will happily update to something more clean.